File: Challenges/checkIfAGivenNodeExistInACompleteTree.java
// https://leetcode.com/discuss/interview-question/236898/Google-Phone-Interview
// https://leetcode.com/discuss/interview-experience/305582/Google-SWE-L3-phone-screen

// Time and space: O(log(n))
public class Main {
    private static final int ROOT = 1;

    public static boolean doesNodeExist(TreeNode root, int target) {
        if (root == null) return false;
        Deque<Integer> path = getPathFromRootTo(target);
        return verifyPath(root, path);
    }

    private static boolean verifyPath(TreeNode node, Deque<Integer> path) {
        while (!path.isEmpty()) {
            node = path.pop() % 2 == 0 ? node.left : node.right;
            if (node == null) return false;
        }
        return true;
    }

    private static Deque<Integer> getPathFromRootTo(int child) {
        Deque<Integer> stack = new ArrayDeque<>();
        while (child != ROOT) {
            stack.push(child);
            child = getParentIndex(child);
        }
        return stack;
    }

    private static int getParentIndex(int childIndex) {
        return childIndex / 2;
    }
    
    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.left.left = new TreeNode(4);
        root.right = new TreeNode(3);
        root.right.left = new TreeNode(6);

        System.out.println(doesNodeExist(root, 5)); // false
        System.out.println(doesNodeExist(root, 7)); // false
        System.out.println(doesNodeExist(root, 20)); // false
        System.out.println(doesNodeExist(null, 1)); // false
        System.out.println(doesNodeExist(root, 1)); // true
        System.out.println(doesNodeExist(root, 2)); // true
        System.out.println(doesNodeExist(root, 3)); // true
        System.out.println(doesNodeExist(root, 4)); // true
        System.out.println(doesNodeExist(root, 6)); // true
    }
}

File: Challenges/checkIfAGivenNodeExistInACompleteTree.py
"""
Given a complete(virtual) binary tree, return true/false if the given target node exists in the tree or not.
Here, the "virtual" means the tree nodes are numbered assuming the tree is a complete binary tree.

For example:

                1
		    /        \ 
		 2              3
       /   \           /  \ 
     4   (5)nil      6  (7)nil
   
   //function signature
   // bool doesNodeExist(root *TreeNode, target int)
   doesNodeExist(root, 4) -> true
   doesNodeExist(root, 7) -> false, Given the node on #7 is a nil node.
   
   //Level order traversal solution, numbering the left(2n+1) and right(2n+2) children node is a trivial solution(time O(n)/O(target) and space O(n)/O(target)). 
   //Think of a better solution.
"""

class TreeNode:
	def __init__(self, val):
		self.val = val
		self.right = None
		self.left = None


def doesNodeExist(root, target):
	if not root:
		return False 

	path = getPathFromRootTo(target)
	return verifyPath(root, path)


def getPathFromRootTo(child):
	stack = []
	while child != 1:
		stack.append(child)
		child = getParentIndex(child)

	return stack


def getParentIndex(childIndex):
	return childIndex//2;


def verifyPath(node, path):
	while path:
		val = path.pop()
		node = node.left if val%2 == 0 else node.right
		if not node:
			return False
	return True

if __name__ == '__main__':
	root = TreeNode(1)
	root.left = TreeNode(2)
	root.left.left = TreeNode(4)
	root.right = TreeNode(3)
	root.right.left = TreeNode(6)
	assert doesNodeExist(root, 5) == False
	assert doesNodeExist(root, 7) == False
	assert doesNodeExist(root, 20) == False
	assert doesNodeExist(None, 1) == False
	assert doesNodeExist(root, 1) == True
	assert doesNodeExist(root, 2) == True
	assert doesNodeExist(root, 3) == True
	assert doesNodeExist(root, 4) == True
	assert doesNodeExist(root, 6) == True

